src/extensions/default/CodeFolding/foldhelpers/foldcode.js

```diff
@@ -0,0 +1,226 @@
+/**
+ * Based on http://codemirror.net/addon/fold/foldcode.js
+ * @author Patrick Oladimeji
+ * @date 10/28/13 8:41:46 AM
+ * @last modified 20 April 2014
+ */
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
+/*global define, brackets, document*/
+define(function (require, exports, module) {
+    "use strict";
+    var CodeMirror          = brackets.getModule("thirdparty/CodeMirror2/lib/codemirror"),
+        prefs               = require("Prefs");
+
+    function doFold(cm, pos, options, force) {
+        options = options || {};
+        force = force || "fold";
+        if (typeof pos === "number") {
+            pos = CodeMirror.Pos(pos, 0);
+        }
+        var finder = options.rangeFinder || CodeMirror.fold.auto;
+        var minSize = options.minFoldSize || prefs.getSetting("minFoldSize");
+
+        function getRange(allowFolded) {
+            var range = options.range || finder(cm, pos);
+            if (!range || range.to.line - range.from.line < minSize) {
+                return null;
+            }
+            var marks = cm.findMarksAt(range.from),
+                i,
+                lastMark,
+                foldMarks;
+            for (i = 0; i < marks.length; ++i) {
+                if (marks[i].__isFold && force !== "fold") {
+                    if (!allowFolded) {
+                        return null;
+                    }
+                    range.cleared = true;
+                    marks[i].clear();
+                }
+            }
+            //check for overlapping folds
+            if (marks && marks.length) {
+                foldMarks = marks.filter(function (d) {
+                    return d.__isFold;
+                });
+                if (foldMarks && foldMarks.length) {
+                    lastMark = foldMarks[foldMarks.length - 1].find();
+                    if (lastMark && range.from.line <= lastMark.to.line && lastMark.to.line < range.to.line) {
+                        return null;
+                    }
+                }
+            }
+            return range;
+        }
+
+        function makeWidget() {
+            var widget = document.createElement("span");
+            widget.className = "CodeMirror-foldmarker";
+            return widget;
+        }
+
+        var range = getRange(true);
+        if (options.scanUp) {
+            while (!range && pos.line > cm.firstLine()) {
+                pos = CodeMirror.Pos(pos.line - 1, 0);
+                range = getRange(false);
+            }
+        }
+        if (!range || range.cleared || force === "unfold" || range.to.line - range.from.line < minSize) {
+            if (range) { range.cleared = false; }
+            return;
+        }
+
+        var myWidget = makeWidget();
```

Comment:
@MiguelCastillo Fwiw, not necessarily true. Because of JSLint, it's safe (often actually saf<i>er</i>) to declare vars further down in a function at the point when they're initialized.  We do have plenty of code in core where vars are not all grouped at the top of the scope.